arcgis-java-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/UtilityNetworkTraceTool.java

```diff
@@ -0,0 +1,999 @@
+
+                UtilityElementTraceResult utilityElementTraceResult =
+                  (UtilityElementTraceResult) utilityTraceResult;
+                // add the element trace results to the current result
+                traceResultInProgress.getElementResults().addAll(
+                  FXCollections.observableArrayList(utilityElementTraceResult.getElements()));
+                traceResultInProgress.setElementResultsByAssetGroup(
+                  traceResultInProgress.getElementResults().stream()
+                    .collect(groupingBy(UtilityElement::getAssetGroup)));
+                // fetch the features to be displayed on the map via an async method
+                fetchFeaturesForElementsFuture =
+                  selectedUtilityNetwork.fetchFeaturesForElementsAsync(utilityElementTraceResult.getElements());
+                fetchFeaturesForElementsFuture.addDoneListener(() -> {
+                  try {
+                    var features = fetchFeaturesForElementsFuture.get();
+                    // add the features to the current result and select the results
+                    traceResultInProgress.getFeatures().addAll(FXCollections.observableArrayList(features));
+                    traceResultInProgress.selectResultFeatures(true);
+                    if (autoZoomToResultsProperty.get()) {
+                      // if authZoomToResult is true, update the viewpoint of the MapView
+                      mapViewProperty.get().setViewpoint(new Viewpoint(traceResultInProgress.getExtent()));
+                    }
+                    // if this is successful, complete the future
+                    fetchFeaturesForElementsCompletableFuture.complete(null);
+                  } catch (CancellationException e) {
+                    // if the async fetch method is cancelled, complete the futures
+                    fetchFeaturesForElementsCompletableFuture.completeExceptionally(e);
+                    traceCompletableFuture.complete(null);
+                  } catch (Exception e) {
+                    // if fetch fails due to another reason, set the error to the result and complete
+                    traceResultInProgress.setException(e);
+                    fetchFeaturesForElementsCompletableFuture.completeExceptionally(e);
+                    traceCompletableFuture.cancel(true);
+                  }
+                });
+              } else if (utilityTraceResult instanceof UtilityGeometryTraceResult) {
+                // handle geometry results and add graphics to the graphics overlay
+                var geometryTraceResult = (UtilityGeometryTraceResult) utilityTraceResult;
+
+                var multipoint = geometryTraceResult.getMultipoint();
+                if (multipoint != null) {
+                  var graphic = new Graphic(multipoint, new SimpleMarkerSymbol(getResultPointSymbol().getStyle(),
+                    getResultPointSymbol().getColor(), getResultPointSymbol().getSize()));
```

Comment:
Have refactored as suggested for adding the graphics.

Regarding the second point, there's quite a few areas where the graphics are referenced - for updating the colors, and also non-MVP functionality which I didn't get time for which was to display the graphics symbols on the results page. I think I was trying to avoid calling `getGraphicsOverlay().getGraphics()` too frequently. But it does feel like 2 sources of truth now you raise it so maybe referencing the graphics via the graphics overlay would be better? Any thoughts?

Reply:
I guess through a `GraphicsOverlay` you can do anything you could with a separate list of `Graphic` so I'd be ok with that - just make it clear in your doc that if you want to get at the graphics go through the provided overlay property.

Reply:
Refactored